### PR TITLE
Remove sig-like type aliases

### PR DESCRIPTION
**Pull request by hovanja2011**

### CODE REVIEW

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 {-# LANGUAGE TypeOperators        #-}
 {-# LANGUAGE UndecidableInstances #-}
 {-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -Wno-type-equality-out-of-scope #-}
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:6]: Better to import `~` from `Data.Type.Equality`

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 2]
```diff
@@ -144,25 +147,19 @@ type family PaddedLength (msg :: Natural) (block :: Natural) (lenBits :: Natural
...
-   , AdditiveSemigroup (UInt (WordSize algorithm) Auto backend)
-   , BoolType (ByteString (WordSize algorithm) backend)
-   , ShiftBits (ByteString (WordSize algorithm) backend)
-   , ShiftBits (ByteString (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) backend)
-   , BoolType (ByteString (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) backend)
-   , Extend (ByteString k backend) (ByteString (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) backend)
-   , ToWords (ByteString (ChunkSize algorithm) backend) (ByteString (WordSize algorithm) backend)
-   , Concat (ByteString (WordSize algorithm) backend) (ByteString (8 * WordSize algorithm) backend)
-   , ToWords (ByteString (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) backend) (ByteString (ChunkSize algorithm) backend)
+   , Extend (ByteString k context) (ByteString (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) context)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:158]: This looks like a redundant constraint

File: examples/Examples/BatchTransfer.hs [hunk 3]
```diff
@@ -1,9 +1,16 @@
...
+import           ZkFold.Symbolic.Cardano.Contracts.BatchTransfer (Tx, TxOut, batchTransfer)
 import           ZkFold.Symbolic.Cardano.Types                   (Bool, ByteString)
+import           ZkFold.Symbolic.Class                           (Symbolic (BaseField))
+import           ZkFold.Symbolic.Data.Eq                         (Eq)
 
 
-exampleBatchTransfer :: (Sig c) => Tx c -> Vector 5 (TxOut c, TxOut c, ByteString 256 c) -> Bool c
+exampleBatchTransfer ::
+    ( Symbolic c
+    , MiMCHash (BaseField c) c (TxOut c, TxOut c)
```
Comment (TurtlePU) [inline examples/Examples/BatchTransfer.hs:13]: This should be inferrable from `Symbolic c`

File: examples/Examples/BatchTransfer.hs [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
+    , Eq (Bool c) (TxOut c)
```
Comment (TurtlePU) [inline examples/Examples/BatchTransfer.hs:14]: This should be inferrable from `Symbolic c`

File: src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs [hunk 5]
```diff
@@ -109,14 +98,15 @@ blake2b_compress Blake2bCtx{h, m, t} lastBlock =
         v1 = V.foldl hashRound v0 $ fromList [0..11] -- twelve rounds
     in fmap (\(i, hi) -> hi `xorUInt` (v1 ! i) `xorUInt` (v1 ! (i GHC.+ 8))) (V.zip (fromList [0..7]) h)
 
-blake2b' :: forall bb' kk' ll' nn' b .
-    ( KnownNat bb'
+blake2b' :: forall bb' kk' ll' nn' c .
+    ( Symbolic c
+    , KnownNat bb'
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs:103]: Are these parameters used on the type level? Maybe we can move them to the value level

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 6]
```diff
@@ -227,30 +221,25 @@ instance (KnownNat n, FromConstant Natural (ByteString n b)) => ToWords Natural
...
-   , KnownNat (WordSize algorithm)
-   , Iso (UInt (WordSize algorithm) Auto backend) (ByteString (WordSize algorithm) backend)
-   , Iso (ByteString (WordSize algorithm) backend) (UInt (WordSize algorithm) Auto backend)
-   , AdditiveSemigroup (UInt (WordSize algorithm) Auto backend)
-   , BoolType (ByteString (WordSize algorithm) backend)
-   , ShiftBits (ByteString (WordSize algorithm) backend)
-   , ToWords (ByteString (ChunkSize algorithm) backend) (ByteString (WordSize algorithm) backend)
-   , Concat (ByteString (WordSize algorithm) backend) (ByteString (8 * WordSize algorithm) backend)
-   )
+type SHA2N algorithm context =
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:224]: Maybe we can move all of these to the superclass constraints of `AlgorithmSetup`? This way, `type SHA2` would also become shorter

File: src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs [hunk 7]
```diff
@@ -29,23 +28,11 @@ type Tx context = Transaction 6 0 11 Tokens 0 () context
...
 hash = mimcHash @(BaseField context) mimcConstants zero
 
-type Sig context =
+verifySignature ::
+    forall context .
     ( Symbolic context
-    , StrictConv (context Par1) (UInt 256 Auto context)
-    , FromConstant Natural (UInt 256 Auto context)
-    , MultiplicativeSemigroup (UInt 256 Auto context)
     , MiMCHash (BaseField context) context (TxOut context, TxOut context)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs:34]: Can this be inferred from `Symbolic context` and, maybe, `FromConstant (BaseField context) (FieldElement context)`?

File: src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs [hunk 8]
```diff
@@ -55,10 +42,11 @@ verifySignature pub (pay, change) sig = (from sig * base) == (strictConv (fromFi
...
 
 batchTransfer ::
-    forall context.  Sig context
-    => Tx context
-    -> Vector 5 (TxOut context, TxOut context, ByteString 256 context)
-    -> Bool context
+    forall context.
+    ( Symbolic context
+    , MiMCHash (BaseField context) context (TxOut context, TxOut context)
+    , Eq (Bool context) (TxOut context)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs:48]: Why is this not inferred from `Symbolic context`?

File: tests/Tests/Blake2b.hs [hunk 9]
```diff
@@ -14,36 +14,25 @@ import           ZkFold.Base.Algebra.Basic.Field             (Zp)
...
 import           ZkFold.Symbolic.Compiler
-import           ZkFold.Symbolic.Data.ByteString             (ByteString, Concat, ReverseEndianness, ShiftBits,
-                                                              ToWords (..), Truncate (..))
+import           ZkFold.Symbolic.Data.ByteString             (ByteString)
 import           ZkFold.Symbolic.Data.Class                  (pieces)
-import           ZkFold.Symbolic.Data.Combinators            (Extend)
 import           ZkFold.Symbolic.Interpreter                 (Interpreter)
 
--- TODO: We need a proper test for both numeric and symbolic blake2b hashing
+-- TODO: We need a proper test for coth numeric and symbolic clake2b hashing
```
Comment (TurtlePU) [inline tests/Tests/Blake2b.hs:23]: typos :)

File: src/ZkFold/Symbolic/Cardano/Contracts/RandomOracle.hs [hunk 10]
```diff
@@ -25,23 +25,11 @@ type Tx context = Transaction 1 0 2 Tokens 1 () context
...
     , FromConstant (BaseField context) (FieldElement context)
-    , MultiplicativeMonoid (UInt 64 Auto context)
-    , Eq (Bool context) (FieldElement context)
-    , Eq (Bool context) (UInt 64 Auto context)
-    , Eq (Bool context) (ByteString 224 context)
-    , Eq (Bool context) (ByteString 256 context)
-    , Extend (ByteString 224 context) (ByteString 256 context)
-    , Extend (AssetName context) (AssetName context)
-    , BinaryExpansion (FieldElement context)
     , Bits (FieldElement context) ~ FieldElementBits context
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Cardano/Contracts/RandomOracle.hs:31]: Maybe it is better to remove `FieldElementBits` alias and this constraint and adapt the contract code for variable number of bits in the basefield? @vlasin, what do you think?

File: src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs [hunk 11]
```diff
@@ -26,26 +29,19 @@ type TxOut context = Output Tokens () context
...
 
 verifySignature ::
-    forall context . Sig context =>
-    ByteString 224 context ->
-    (TxOut context, TxOut context) ->
-    ByteString 256 context ->
-    Bool context
+    forall context .
+    ( Symbolic context
+    , SymbolicData (TxOut context, TxOut context)
```
Comment (vlasin) [inline src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs:43]: Can we derive it from `SymbolicData (TxOut context)`?

**Review by TurtlePU — APPROVED** [review]

File: src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs (hunk 11 above)
Comment (TurtlePU) [reply src/ZkFold/Symbolic/Cardano/Contracts/BatchTransfer.hs:43]: This also needs `KnownNat (TxOut context)` until we refactor `SymbolicData`

**Review by TurtlePU — APPROVED** [review]

File: src/ZkFold/Symbolic/Cardano/Contracts/RandomOracle.hs (hunk 10 above)
Comment (TurtlePU) [reply src/ZkFold/Symbolic/Cardano/Contracts/RandomOracle.hs]: @vlasin, maybe we can do it in a separate PR then? This is not in scope of this PR IMO

**Review by vlasin — APPROVED** [review]

**Review by vks4git — APPROVED** [review]

**Pull request merged**